### DISCUSSION TITLE
Why isn't meep designed to be thread-safe?

### DISCUSSION THREAD

**Discussion by mountqi** [NanoComp/meep#1932] [Q&A]
There're many global variables in the source code. If called in multi-threads, meep may crash.

I want to know why is it designed so?

**Comment by oskooi** [comment]
Meep itself is generally thread safe and in fact supports [thread-level parallelism via OpenMP](https://meep.readthedocs.io/en/latest/Parallel_Meep/#different-forms-of-parallelization).  See #1628. (There are very few instances of global state variables in https://github.com/NanoComp/meep/blob/master/src/meep.hpp.)

One of Meep's dependencies, [libctl](https://github.com/NanoComp/libctl), does use a global variable to specify the [`default_material`](https://github.com/NanoComp/meep/blob/61fae689eb7d891223bbc9d4f24827d196938d51/src/meepgeom.cpp#L26-L51) and it would be good to change this instead to an argument of all functions which depend on it so that they are reentrant. However, since the default material is almost always time independent and thus never modified during the time stepping, this has not been an issue.
